### PR TITLE
[mxfp8 moe training] add cuda kernel for per group padding

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Stacked PRs:
 * #4021
 * __->__#3998

--- --- ---

## Motivation
- Torchtitan is removing the token group padding logic since it is now only needed for fp8/mxfp8, not bf16 grouped mm
- We have also received [user feedback](https://github.com/pytorch/ao/issues/3287#issuecomment-3493036157) that torchao should handle this token group padding / data movement as it is tricky to do efficiently and many users won't be able to get a speedup using TorchAO MXFP8 without an optimized implementation of it.

## Summary
- For emulated mode, use a torch native token group padding impl with d2h sync
- For non-emulated mode, I prototyped the following:
    - Add 3 prototype kernels to do this padding with different approaches, with tests and benchmarks
       - Approach 1: Triton two stage approach;
           -  Add padding row to inputs with torch.vstack
           - Triton kernel generates indexes with padding index of -1, avoiding d2h sync
           - Torch code uses those indexes like `inputs[padded_indexes, :]` where the -1 index will select a padding row every time it appears.
       - Approach 2: Triton fused approach (do copy tokens and write padding data in one kernel )
       - Approach 3: CUDA kernel with 2 stage approach (precompute padded offsets in tiny kernel instead of several torch ops, then dispatch padding kernel)

## Tests
- `pytest test/prototype/moe_training/test_kernels.py -k pad`

## Benchmarks
 CUDA is the best across all shapes:
```
  num_tokens    dim    num_groups    torch_us    triton_us    fused_triton_us    cuda_us    torch_mem_bw_gbps    triton_mem_bw_gbps    triton_fused_mem_bw_gbps    cuda_mem_bw_gbps  triton_vs_torch    fused_triton_vs_torch    cuda_vs_torch
------------  -----  ------------  ----------  -----------  -----------------  ---------  -------------------  --------------------  --------------------------  ------------------  -----------------  -----------------------  ---------------
       16384   1536             1     436.458      439.953            417.303    320.444               346.18                343.43                      362.07              471.51  0.99x              1.05x                    1.36x
       16384   1536             4     517.7        452.393            383.254    292.423               292.42                334.64                      395.01              517.7   1.14x              1.35x                    1.77x
       16384   1536             8     536.612      466.062            368.165    286.548               282.85                325.67                      412.27              529.69  1.15x              1.46x                    1.87x
       16384   1536            16     598          465.822            367.509    286.343               255.13                327.52                      415.14              532.81  1.28x              1.63x                    2.09x
       16384   2048             1     441.902      449.814            382.631    292.373               455.89                447.87                      526.51              689.04  0.98x              1.15x                    1.51x
       16384   2048             4     484.005      461.893            401.875    309.171               417.04                437.01                      502.27              652.88  1.05x              1.20x                    1.57x
       16384   2048             8     541.357      473.953            392.233    298.03                373.83                426.99                      515.96              679.04  1.14x              1.38x                    1.82x
       16384   2048            16     585.292      460.7              390.033    293.395               347.56                441.55                      521.56              693.34  1.27x              1.50x                    1.99x
       16384   5120             1     460.036      464.683            567.816    299.987              1094.79               1083.85                      886.98             1678.89  0.99x              0.81x                    1.53x
       16384   5120             4     497.396      449.034            575.571    286.81               1014.54               1123.81                      876.74             1759.45  1.11x              0.86x                    1.73x
       16384   5120             8     526.011      455.096            612.969    300.776               961.84               1111.72                      825.39             1682.11  1.16x              0.86x                    1.75x
       16384   5120            16     607.179      484.409            630.304    296.039               837.58               1049.86                      806.85             1717.88  1.25x              0.96x                    2.05x
       16384   7168             1     465.052      459.34             613.543    295.84               1516.18               1535.03                     1149.23             2383.39  1.01x              0.76x                    1.57x
       16384   7168             4     492.211      456.083            623.67     293.7                1435.31               1549.01                     1132.77             2405.44  1.08x              0.79x                    1.68x
       16384   7168             8     533.49       465.685            665.016    293.333              1327.7                1521.01                     1065.11             2414.71  1.15x              0.80x                    1.82x
       16384   7168            16     609.1        469.449            677.495    298.956              1168.91               1516.64                     1050.91             2381.56  1.30x              0.90x                    2.04x
```
## Limitations
- We should update the docs to recommend Torchtitan users use HybridEP which fuses the padding into the dispatch step to avoid this extra copy (which is expensive since input activations are huge).
- For non-torchtitan users, we can modify our [Triton+Symmetric memory all2all](https://github.com/pytorch/ao/blob/d6d423e0322b760cf8c5d69934915a837f319a99/torchao/prototype/moe_training/kernels/mxfp8/comms.py#L25) impl to fuse padding this way as well, but it is a larger project.

## MXFP8 grouped mm autograd func fwd + bwd new benchmarks

We lose about 25% of the speedup for Llama4 and 50% of the speedup for DSV3 671b. Need to do the all2all dispatch padding approach to avoid this.

```
M,N,K,G                  recipe                             bf16_fwd_bwd_us    scaled_fwd_bwd_us  scaled_fwd_bwd_speedup      bf16_fwd_us    scaled_fwd_us  scaled_fwd_speedup
-----------------------  -------------------------------  -----------------  -------------------  ------------------------  -------------  ---------------  --------------------
(32768, 8192, 5120, 1)   MXFP8TrainingRecipe.MXFP8_RCEIL            7274.69              4548.62  1.599x                         2033.6           1249.18   1.628x
(32768, 8192, 5120, 2)   MXFP8TrainingRecipe.MXFP8_RCEIL            7264.42              4645.89  1.564x                         2171.97          1318.75   1.647x
(128000, 8192, 5120, 1)  MXFP8TrainingRecipe.MXFP8_RCEIL           28025.1              17680.4   1.585x                         9683.04          4871.17   1.988x
(128000, 8192, 5120, 2)  MXFP8TrainingRecipe.MXFP8_RCEIL           27889.7              17515.6   1.592x                         9126.48          4927.33   1.852x
(32768, 2048, 7168, 4)   MXFP8TrainingRecipe.MXFP8_RCEIL            2493.47              2234.4   1.116x                          733.216          758.752  0.966x
(32768, 2048, 7168, 8)   MXFP8TrainingRecipe.MXFP8_RCEIL            2519.04              2359.2   1.068x                          708.544          807.872  0.877x
(128000, 2048, 7168, 4)  MXFP8TrainingRecipe.MXFP8_RCEIL           10091.2               8602.66  1.173x                         2898.9           2812.06   1.031x
(128000, 2048, 7168, 8)  MXFP8TrainingRecipe.MXFP8_RCEIL            9958.98              8097.26  1.23x                          3020.8           2861.15   1.056x
```

### CODE REVIEW

File: torchao/prototype/moe_training/mxfp8_grouped_mm.py [hunk 1]
```diff
@@ -716,5 +736,32 @@ def _emulated_mxfp8_scaled_grouped_mm_2d_2d(
     return out
 
 
+def _groups_aligned(
+    group_offsets: torch.Tensor, alignment_size: int = 32
+) -> torch.Tensor:
+    group_sizes = torch.diff(
```
Comment (drisspg) [inline torchao/prototype/moe_training/mxfp8_grouped_mm.py]: thats a fun one claude, instead of cumusm no weird sync things here right?

File: torchao/prototype/moe_training/mxfp8_grouped_mm.py [hunk 2]
```diff
@@ -716,5 +736,32 @@ def _emulated_mxfp8_scaled_grouped_mm_2d_2d(
...
+        group_offsets, prepend=torch.zeros((1,), device=group_offsets.device)
+    )
+    groups_aligned = ((group_sizes % alignment_size) == 0).all()
```
Comment (drisspg) [inline torchao/prototype/moe_training/mxfp8_grouped_mm.py]: well here is a sync which you mgith ahve said is expected?

File: torchao/prototype/moe_training/kernels/mxfp8/quant.py [hunk 3]
```diff
@@ -245,11 +246,383 @@ def compute_blocked_scale_offsets_for_K_groups(
...
+        padded_group_offsets: New group offsets after padding
+    """
+    num_tokens, dim = inputs.shape
+
+    padded_groups = []
+    padded_offsets = []
+    current_offset = 0
+    group_start = 0
+
+    for group_end in group_offsets.tolist():
```
Comment (drisspg) [inline torchao/prototype/moe_training/kernels/mxfp8/quant.py]: use @howardzhang-cv skill and see if we can also try helion here

Comment (danielvegamyhre) [reply torchao/prototype/moe_training/kernels/mxfp8/quant.py]: could be a great starter task for new team member who starts Monday, they're going to work on some training stuff and apparently they are interested in helion

File: torchao/prototype/moe_training/mxfp8_grouped_mm.py (hunk 1 above)
Comment (danielvegamyhre) [reply torchao/prototype/moe_training/mxfp8_grouped_mm.py]: deleted this and made more compile friendly by just using a `pad_token_groups: bool` arg in the autograd func, for the caller to specify if they need their token groups aligned or not. 

was having a bad time with data dependent control flow issues, etc

File: torchao/prototype/moe_training/mxfp8_grouped_mm.py (hunk 2 above)
Comment (danielvegamyhre) [reply torchao/prototype/moe_training/mxfp8_grouped_mm.py]: deleted this, see other comment
